landlab/components/groundwater/dupuit_percolator.py

```diff
@@ -752,6 +753,6 @@ def run_with_adaptive_time_step_solver(self, dt):
             remaining_time -= substep_dt
             self._num_substeps += 1
 
-            self._callback_fun(self._grid, substep_dt)
+            self._callback_fun(self._grid, self.recharge, substep_dt)
```

Comment:
For backward compatibility, I think that this needs to take either 2 or 3 arguments. 

1. The docs should state this, and that the two option callback function is deprecated. 

2. In the code itself the number of arguments should get tested. One example of testing the number of arguments a function takes is given at line 168 of landlab\components\lithology\litholayers.py @mcflugen may know a better way to do this. 

3. If two arguments are provided then a DeprecationWarning should be raised and grid, dt passed. If three, then grid, recharge, dt passed. (Sidenote: Raising the DeprecationWarning will be helpful for whoever manages the next major version because it will help them find that this is a change that was deprecated). 

4. Add one test, so you have one test with a three element function and one with a two element function (that yields a deprecation warning).  At some point I remember it being tricky to figure out how to test for deprecated warnings. Here is a link [to the part of pytest](https://docs.pytest.org/en/stable/warnings.html#ensuring-code-triggers-a-deprecation-warning) that shows an example.

Reply:
Thanks for taking a look Katy. I'm trying out the method `function.__code__.co_varnames`, and the one problem I see is that it doesn't distinguish between arguments and keyword arguments. There can be as many keyword arguments as the user wants, e.g. to supply the name of the output file. I'm looking into using the `signature` class from inspect, but I'm still not quite sure how to separate the two. For example, someone could provide a function something like:

```
def callback(grid, recharge, dt, d='file.csv'):
    return np.save_txt(d, grid['surface_water_specific__discharge']*dt)
```
And that would be fine.